File: task4/src/main/java/Algorithms/QuickSort.java
package Algorithms;

import java.util.Arrays;

public class QuickSort {
    public static int partition(int[] arr, int l, int r){
        int pivot = arr[(l + r) / 2];
        int leftMove = l;
        int rightMove = r;
        while (leftMove <= rightMove) {
            while (arr[leftMove] < pivot) {
                leftMove++;
            }
            while (arr[rightMove] > pivot) {
                rightMove--;
            }
            if (leftMove >= rightMove) {
                break;
            }
            int temp = arr[leftMove];
            arr[leftMove] = arr[rightMove];
            arr[rightMove] = temp;
            leftMove++;
            rightMove--;
        }
        return rightMove;
    }
    public static void ExecuteSort(int[] arr, int l, int r){
        if (l < r){
            int q = partition(arr, l, r);
            ExecuteSort(arr, l, q);
            ExecuteSort(arr, q + 1, r);
        }
    }
    public static void main(String[] args) {
        int[] arr = new int[10000];
        int[] ar2 = new int[10000];
        for (int i = 0; i < arr.length; i++) {
            arr[i] = (int) (Math.random() * 1000);
            ar2[i] = arr[i];
        }
        ExecuteSort(arr, 0, arr.length - 1);

        Arrays.sort(ar2);
        System.out.println(Arrays.equals(arr, ar2));
    }
}

File: lab-1/task4/QuickSort.cs
using System;

namespace task4
{
    public class QuickSort
    {
        public static int[] CreateRandomArr(int count)
        {
            var arr = new int[count];
            var random = new Random();
            for (int i = 0; i < arr.Length; i++)
            {
                arr[i] = random.Next(0, Int32.MaxValue);
            }

            return arr;
        }

        private static int Partition(int[] arr, int l, int r)
        {
            int pivot = arr[(l + r) / 2];
            int leftMove = l;
            int rightMove = r;
            while (leftMove <= rightMove)
            {
                while (arr[leftMove] < pivot)
                {
                    leftMove++;
                }

                while (arr[rightMove] > pivot)
                {
                    rightMove--;
                }

                if (leftMove >= rightMove)
                {
                    break;
                }

                (arr[leftMove], arr[rightMove]) = (arr[rightMove], arr[leftMove]);
                leftMove++;
                rightMove--;
            }

            return rightMove;
        }

        public static void ExecuteSort(int[] arr, int l, int r)
        {
            if (l < r)
            {
                int q = Partition(arr, l, r);
                ExecuteSort(arr, l, q);
                ExecuteSort(arr, q + 1, r);
            }
        }

        public static void StartExecute(int count)
        {
            var arr = CreateRandomArr(count);
            ExecuteSort(arr, 0, arr.Length - 1);
        }
    }
}